core/engine/linksystem.go

```diff
@@ -0,0 +1,89 @@
+package engine
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/filecoin-project/indexer-reference-provider/core"
+	schema "github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+const (
+	advPrefix      = "adv/"
+	indexPrefix    = "index/"
+	latestAdvKey   = "sync/adv"
+	latestIndexKey = "sync/index"
+)
+
+// LinkSystem for the reference provider
+func mkLinkSystem(ds datastore.Batching) ipld.LinkSystem {
+	lsys := cidlink.DefaultLinkSystem()
+	lsys.StorageReadOpener = func(lctx ipld.LinkContext, lnk ipld.Link) (io.Reader, error) {
+		val, err := ds.Get(datastore.NewKey(storageKey(lctx, lnk)))
+		if err != nil {
+			return nil, err
+		}
+		return bytes.NewBuffer(val), nil
+	}
+	lsys.StorageWriteOpener = func(lctx ipld.LinkContext) (io.Writer, ipld.BlockWriteCommitter, error) {
+		buf := bytes.NewBuffer(nil)
+		return buf, func(lnk ipld.Link) error {
+			return ds.Put(datastore.NewKey(storageKey(lctx, lnk)), buf.Bytes())
+		}, nil
+	}
```

Comment:
@warpfork consider if there's a way to avoid this boilerplate?

Reply:
I don't think we can (or want) to avoid this boilerplate. Each link system may implement a different interaction with the underlying datastore. In this case, we add a small prefix according to the schema type to identify to which schema type a CID belongs to.